webapp/src/components/Analysis/PerturbationTestingPreview.test.tsx

```diff
@@ -0,0 +1,167 @@
+
+  it("should have two toggle buttons for the train and test", async () => {
+    renderPerturbationTestingPreview({ train: true, eval: true });
+    await waitFor(() => {
+      expect(screen.getAllByRole("button")).toHaveLength(2);
+    });
+  });
+
+  it("should have a disabled toggle button if either one of the dataset split is unavailable", async () => {
+    renderPerturbationTestingPreview({ train: false, eval: true });
+    await waitFor(() => {
+      //Verifying if the train toggle button is disabled if not available in datasplits
+      expect(screen.getByRole("button", { pressed: false })).toHaveClass(
+        "Mui-disabled"
+      );
+      // By default, evaluation toggle is selected on page load.
+      expect(screen.getByRole("button", { pressed: true })).not.toHaveClass(
+        "Mui-disabled"
+      );
+    });
+  });
+
+  it("should display the data in the toggle with expected styling (color, font), format", async () => {
+    renderPerturbationTestingPreview({ train: true, eval: true });
+    await waitFor(() => {
+      // Verify if the actual response data from API is visibly displayed on the toggle
+      expect(screen.getByText(/14.0%/i)).toBeVisible();
+      expect(screen.getByText("Failure rate - Evaluation Set")).toBeVisible();
+      expect(screen.getByText(/16.0%/i)).toBeVisible();
+      expect(screen.getByText("Failure rate - Training Set")).toBeVisible();
+
+      // styling should be applied for the selected toggle as expected
+      expect(screen.getByText(/14.0%/)).toHaveStyle(
+        "font-size: 3.75rem; color: rgb(237, 108, 2)"
+      );
+
+      // verify with some random unexpected values to not have been displayed.
+      expect(screen.queryByText(/17.0%/i)).toBeNull();
+      expect(screen.queryByText("some random text")).toBeNull();
+      expect(screen.queryByText(/26.0%/i)).toBeNull();
+    });
+  });
+
+  it("should not display a value if datasplit is not available", async () => {
+    renderPerturbationTestingPreview({ train: false, eval: true });
+    await waitFor(() => expect(screen.getByText("--%")).toBeVisible());
+  });
+
+  it("should have expected columns", async () => {
+    renderPerturbationTestingPreview({ train: true, eval: true });
+    const expectedColumnHeaders = [
+      "Test Family",
+      "Test Name",
+      "Modif. Type",
+      "Test Description",
+      "FR on Evaluation Set",
+    ];
+    const actualColumnHeaders = screen.getAllByRole("columnheader");
+    await waitFor(() => {
+      expectedColumnHeaders.forEach((name, index) => {
+        expect(actualColumnHeaders[index].textContent).toEqual(name);
+      });
+    });
+  });
+
+  it("should have default sort as desc for colum header 'FR on Evaluation/Training set'", async () => {
+    renderPerturbationTestingPreview({ train: true, eval: true });
+    await waitFor(() =>
+      expect(
+        screen.getAllByRole("columnheader")[4].getAttribute("aria-sort")
+      ).toBe("descending")
+    );
```

Comment:
3 things:
- `waitFor` is not necessary as this is rendered even before the data comes in;
- we can get the Failure Rate column directly instead of hardcoding that it is the 5th one; and
- we can check for attribute directly, to be more straightforward.
```diff
-expect(<...>.getAttribute("aria-sort")).toBe("descending")
+expect(<...>).toHaveAttribute("aria-sort", "descending")
```
With everything together:
```suggestion
    expect(
      screen.getByRole("columnheader", { name: /Failure Rate/ })
    ).toHaveAttribute("aria-sort", "descending");
```

Reply:
Understood. I can make use of aria-label as name in this cases. Good point to be followed for future tests

Reply:
In some rare scenario, I have got undefined error so added "waitFor" here.